### PR TITLE
2022 December Build

### PR DESCRIPTION
**Pull request by YuelinXin**
# 2022 December Build

## What's New?
In the past few months, we've been working on a new version of our website. We've been thinking of a new design, new features, and a new way to use our website. And here it is!

First of all, we've added a sign up & sign in function. Now you can create your own MF account and log into it to save your community settings, the projects you are involved in, the teams you are a part of, and so on. These are all shown elegantly on your profile page and we are adding ways for you to edit / delete them more easily. We are also working on a way to let you sign up / sign in with your GitHub account so that you won't have to sign in with your password every time.

Secondly, we've added a new *People* page. You can now see all the amazing people involved in our community. You can also contact them by clicking the email link beside their names. We are also working on a public version of everyone's profile page so that you can check out what others' been up to and interact with them more intuitively.

## What's Changed? (code-wise)
Through the work that we have been doing for the past few months, we made our website even more integrated with Flask and Flask packages. We've added Flask forms, template inheritance, template macros, error / debug log handling and so on. 

We've also made our codebase more professional by integrating unit testing, database migration, and other features that are commonly used (well, I do have to admit that our testing is still pretty basic and stupid at the moment). We've also tried to organize our codebase better by separating the code into different modules and packages for better readability and maintainability.

We have also made our website more secure by hashing user passwords and checking for signs of SQL injection to protect the users' data from malicious attacks.

## What's Next?
We understand that this build is not perfect yet, and we know very well that some key aspects of this project are still in very poor shape, but the development team will actively work on our website and make it more usable, filled with more functionalities, more security checks, and more elegant styling.

Here are some basic ideas and aspects that we are currently working on:
- Adding a way to sign up / sign in with GitHub account
- Adding a way to edit / delete your projects / teams
- Building a public version of everyone's profile page
- Building the *Projects*, *Teams*, *Research*, and *Campus* pages
- Adding some cool features so that people can communicate with each other more easily
- Adding a way to search for information on the website
- Moving into a more professional and reliable hosting service
- Moving into a more professional CI/CD pipeline
- Refine our unit testing
- Adding more documentation so other people can contribute to our project more easily
- Tackle more security issues

Our website is still lacking key functionalities that makes it usable, we are working on tackling these tasks tirelessly and we hope that we can make our website more usable and more useful for everyone.

### CODE REVIEW

File: application/__init__.py [hunk 1]
```diff
@@ -11,37 +11,105 @@
...
+
+    # database configuration
+    app.config['SQLALCHEMY_DATABASE_URI'] = config('database_uri', '')
+    app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False
+
+    # save uploaded files configuration
+    app.config['UPLOAD_FOLDER'] = os.path.join('.', os.path.dirname(__file__), 
+                                               config('upload_folder', ''))
+
+    # sijax configuration
```
Comment (YuelinXin) [inline application/__init__.py:75]: We will add sijax functions in the future

File: application/admin/models.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
+# ///////////////////////////////////////////////////////////////////////////
+# @file: admin/models.py
```
Comment (YuelinXin) [inline application/admin/models.py:2]: This file is not being used at the moment

File: application/email.py [hunk 3]
```diff
@@ -14,21 +14,37 @@
...
 
 
 # generate email
-def send_email(user, email):
+def send_email(user, email, type):
     # send confirmation email
     token = generate_confirmation_token(email)
-    # email_link = 'https://miraclefactory.co/confirm/' + token   # deploy
-    email_link = 'http://localhost:9000/confirm/' + token   # debug
+    email_link = config('email_prefix', '')+'confirm/'+token+'/'+type
```
Comment (YuelinXin) [inline application/email.py:28]: This is added so that we won't need to switch between the deploy and debug version

File: application/forms.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
+# ///////////////////////////////////////////////////////////////////////////
+# @file: forms.py
```
Comment (YuelinXin) [inline application/forms.py:2]: This file still needs a lot more work to make the code look better

File: application/user/models.py [hunk 5]
```diff
@@ -0,0 +1,235 @@
+# ///////////////////////////////////////////////////////////////////////////
+# @file: user/models.py
```
Comment (YuelinXin) [inline application/user/models.py:2]: This file can still use some work

File: application/views.py [hunk 6]
```diff
@@ -9,20 +9,47 @@
```
Comment (YuelinXin) [inline application/views.py:9]: This file can potentially be separated into several different files

**Review by YuelinXin — COMMENTED** [review]
2022 December Build is AOK

File: application/__init__.py (hunk 1 above)
Comment (zzh8241102) [reply application/__init__.py:75]: lgtm

**Comment by zzh8241102** [comment]
look good to me

**Review by zzh8241102 — APPROVED** [review]

**Pull request merged**